base/tuple_from_iterator.jl

```diff
@@ -0,0 +1,169 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module TupleFromIterator
+
+import ..TypeIntersectExact as TIE
+
+incremented(n::Int) = Core.Intrinsics.add_int(n, 1)
+decremented(n::Int) = Core.Intrinsics.sub_int(n, 1)
+
+function _totuple_err(@nospecialize T)
+    @noinline
+    throw(ArgumentError("too few or too many elements for tuple type $T"))
+end
+
+function iterator_to_ntuple_recur(::Type{Tuple{}}, iter, state::Union{Tuple{},Tuple{Any}})
+    i = iterate(iter, state...)::Union{Nothing,Tuple{Any,Any}}
+    isnothing(i) ||
+        throw(ArgumentError("iterator has too many elements for the tuple type"))
+    ()
+end
+function iterator_to_ntuple_recur(
+    ::Type{Tuple{E,Vararg{E,lenm1}}}, iter, state::Union{Tuple{},Tuple{Any}}
+) where {E,lenm1}
+    T2 = Tuple{E,Any}
+    ItUn = Union{Nothing,T2}
+    i = iterate(iter, state...)::ItUn
+    T = Tuple{E,Vararg{E,lenm1}}
+    isnothing(i) &&
+        throw(ArgumentError("iterator has too few elements for the tuple type"))
+    (e, s) = i::T2
+    Next = NTuple{lenm1,E}
+    t = iterator_to_ntuple_recur(Next, iter, (s,))::Next
+    (e, t...)::T
+end
+
+function iterator_to_ntuple(::Type{T}, iter) where {len,T<:NTuple{len,Any}}
+    if len < 100
```

Comment:
100 seems pretty high for this. Does the recursive version really infer that well?